### PR TITLE
Minimum version checking refactor

### PR DESCRIPTION
**Pull request by mrengstrom**
Background:
Spawned from a discussion in slack about agent compilation errors when the installed go version is too old.  General idea is to be able to setup a set of version dependencies for caldera and be able to evaluate them and notify the user on server boot.

### CODE REVIEW

File: app/utility/base_world.py [hunk 1]
```diff
@@ -109,6 +112,26 @@ async def walk_file_path(path, target):
...
+    @staticmethod
+    def check_requirement(params):
+        def parse_version(version_string, pattern=r'([0-9]+(?:\.[0-9]+)+)'):
+            groups = re.search(pattern, version_string)
+            if groups:
+                return groups[1]
+            return '0.0.0'
+
+        try:
+            if 'module' in params:
```
Comment (privateducky) [inline app/utility/base_world.py:124]: ugly function warning ;)

File: conf/default.yml [hunk 2]
```diff
@@ -20,6 +20,14 @@ plugins:
 - access
 port: 8888
 reports_dir: /tmp
+requirements:
+  go:
+    command: go version
+    version: 1.11
```
Comment (privateducky) [inline conf/default.yml:26]: great spot for this

File: server.py [hunk 3]
```diff
@@ -48,6 +48,7 @@ async def start_server():
 def run_tasks(services):
     loop = asyncio.get_event_loop()
     loop.create_task(build_docs())
+    loop.run_until_complete(app_svc.validate_requirements())
```
Comment (privateducky) [inline server.py:51]: this can likely be a create_task instead of run_until_complete

File: app/utility/base_world.py (hunk 1 above)
Comment (mrengstrom) [reply app/utility/base_world.py]: hopefully the villagers have not gathered their pitchforks yet, I just pushed a refactor to compartmentalize things and split out functionality

File: app/utility/base_world.py [hunk 4]
```diff
@@ -111,6 +113,39 @@ async def walk_file_path(path, target):
...
+            'python_module': check_module_version,
+            'installed_program': check_program_version
+        }
+
+        try:
+            requirement_type = params.get('type')
+            return checkers[requirement_type](**params)
+        except Exception as e:
+            logging.getLogger('check_requirement').error(repr(e))
+        return False
```
Comment (privateducky) [inline app/utility/base_world.py:147]: remove return, not used

File: app/utility/base_world.py [hunk 5]
```diff
@@ -111,6 +113,39 @@ async def walk_file_path(path, target):
...
+            version = parse_version(version_string)
+            return distutils.version.StrictVersion(version) >= distutils.version.StrictVersion(str(minimum_version))
+
+        def parse_version(version_string, pattern=r'([0-9]+(?:\.[0-9]+)+)'):
+            groups = re.search(pattern, version_string)
+            if groups:
+                return groups[1]
+            return '0.0.0'
+
+        checkers = {
```
Comment (privateducky) [inline app/utility/base_world.py:137]: swap to dict() format instead of {}

File: app/utility/base_world.py [hunk 6]
```diff
@@ -111,6 +113,39 @@ async def walk_file_path(path, target):
                 return os.path.join(root, '%s.xored' % target)
         return None
 
+    @staticmethod
+    def check_requirement(params):
+        def check_module_version(module, version, attr=None, **kwargs):
```
Comment (privateducky) [inline app/utility/base_world.py:118]: this is nice

**Comment by wbooth** [comment]
looks clean

**Pull request merged**